### PR TITLE
Added model and API to subscribe to comments and posts

### LINKED ISSUE #465: Mail: model for tracking post/comment subscriptions

### LINKED ISSUE #466: Mail: automatically subscribe users to their own posts and comments

### LINKED ISSUE #468: Mail: API for follow / unfollow
We need a simple python API to subscribe users to their own posts and comments. 

We also need a REST API to expose this to the front-end, in support of #469.

### PR DESCRIPTION
**Pull request by rhysyngsun**
#### What are the relevant tickets?
Fixes #466 
Fixes #468 
Fixes #465 

#### What's this PR do?
Adds model and API to add/remove post/comment subscriptions

#### How should this be manually tested?
Go to the APIs endpoints for a post and comment (`/api/v0/post/ID/` and `/api/v0/comment/ID` respectively) and verify you can `PATCH` those with `{"subscribed": true|false}` and the desired behavior happens and a corresponding `Subscription` record is created or deleted as appropriate.

#### Any background context you want to provide?
I implemented a `Base36IntegerField` that we can use going forward when we need to store a reddit id somewhere. It accepts base-36 strings and returns them while storing them in a bigint field the same as the reddit backend does. This make the data type opaque from an application perspective but gives us a backend representation that aligns with reddit's. This was an easier approach than manually encoding/decoding wherever was needed.

### CODE REVIEW

**Comment by aliceriot** [comment]
I'm not sure if this is the intended behavior, but I just ran into something a little puzzling. I went to check the API, and, before checking, I ran a query to see if there were any `Subscription` records already existing for my post. I found two, both of which had `comment_id`'s for the two comments I made on the post. The post itself didn't have a record.

The API for the Post was returning `subscribed: True`.

Then when I patched `{ "subscribed": false }` on the Post API, it ended up deleting these records, and now the API returns `subscribed: False`. Now if I patch with `{ "subscribed": true }` it creates a Post subscription (i.e. without a `comment_id` value), and patching again with `{ "subscribed": false }` deletes the record.

My question is, should patching `{ "subscribed": false }` on the Post API delete the comment subscriptions too?

**Comment by aliceriot** [comment]
the comment API itself works as I would expect, I can add and remove subscriptions on individual comments without them affecting each other.

**Comment by rhysyngsun** [comment]
Ahhhhh, good catch, the API needs to filter to `comment_id__isnull=True` when operating on post subscriptions!

**Comment by rhysyngsun** [comment]
In my testing I also found that if you had both a post and comment subscription and tried to create a new post one, it would error on the `get_or_create` since the `get` piece was returning two records.

**Comment by rhysyngsun** [comment]
@aliceriot all set

**Comment by rhysyngsun** [comment]
I also just added an extra index I realized I'm going to need downstream when querying for subscriptions  without a user.

**Comment by aliceriot** [comment]
cool, just checked the functionality and it looks great! reading code now

File: channels/serializers.py [hunk 1]
```diff
@@ -218,6 +220,20 @@ def validate_url(self, value):
...
+        """Validate that subscribed is a bool"""
+        return {'subscribed': _parse_bool(value, 'subscribed')}
+
+    def get_subscribed(self, instance):
+        """Returns True if user is subscribed to the post"""
+        if 'post_subscriptions' not in self.context:
+            # this code is run if a post was just created
+            return Subscription.objects.filter(
+                user=self.context['request'].user,
+                post_id=instance.id,
```
[the remark this answers is not in the document]
Comment (rhysyngsun) [reply channels/serializers.py]: I'll do it anyway to be safe.

**Comment by rhysyngsun** [comment]
@aliceriot feedback addressed

**Pull request merged**